core/src/mindustry/ui/dialogs/ResearchDialog.java

```diff
@@ -792,4 +859,43 @@ public void drawChildren(){
             super.drawChildren();
         }
     }
-}
+    //Whether this player is allowed to research (for use in multiplayer campaign)
+    public boolean researchAllowed(@Nullable Player player){
+        return !Vars.net.active() || (player != null && (player.admin() || !state.rules.researchRequiresAdmin));
+    }
+    // TODO shouldn't this have a target?
+    @Remote
+    public static void partiallyResearched(Content content, int[] ids, int[] quantity){
+        if(!(content instanceof UnlockableContent u) || u.techNode == null) return;
+        if(net.client()) { // TODO isn't this always true?
```

Comment:
```suggestion
        if(net.server(() {
```
I presume the server is the one researching... why would you want the people playing to handle this?